pkg/crypto/certloader/client.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2020 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package certloader
+
+import (
+	"crypto/tls"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+)
+
+// ClientConfigBuilder creates tls.Config to be used as TLS client.
+type ClientConfigBuilder interface {
+	IsMutualTLS() bool
+	ClientConfig(base *tls.Config) *tls.Config
+}
+
+// WatchedClientConfig is a ClientConfigBuilder backed up by files to be
+// watched for changes.
+type WatchedClientConfig struct {
+	*Watcher
+	log logrus.FieldLogger
+}
+
+// NewWatchedClientConfig returns a WatchedClientConfig configured with the
+// provided files. When caFiles is nil or empty, the system CA CertPool is
+// used. To configure a mTLS capable ClientConfigBuilder, both certFile and
+// privkeyFile must be provided.
```

Comment:
So you are correct in that the client doesn't use the system cert pool directly.

I've removed the loading of `x509.SystemCertPool()` when I realized that when `tls.Config` has its `RootCAs` member set to `nil` then the system cert pool is used, see https://golang.org/pkg/crypto/tls/#Config. So effectively we don't need to load the system cert pool but it will still be usedwhen `caFiles` is empty.